### PR TITLE
Finalize container value extractors

### PR DESCRIPTION
**Pull request by yrodiere**
Well, maybe it's not "finalized", but at least it's not just a hack anymore.

Main changes:

 * Improved generic type introspection (the part where we get the type arguments of a parameterized type, for instance) greatly. The previous implementation was just a hack and was very fragile.
 * Added more built-in extractors (Optionals, generic arrays, map keys)
 * Added support for recursive container value extractor resolution (e.g. `Map<String, List<SomeEntity>>`). Of course it's not very useful in the ORM integration, since ORM doesn't support those types very well, but it might be useful in non-ORM uses of Hibernate Search.
 * Added ways to configure the extractors to use in an `@IndexedEmbedded`/`@Field` explicitly
 * Got rid of hard-coded type matching, adding the capability to infer supported type from the type arguments to the `ContainerValueExtractor` interface in the implementation type. This effectively allows users to implement their own extractor.
 * Cleaned up the code, splitting it into three parts: type introspection (package `.model.spi`), type matching (`.model.typematching.impl`) and container creation/resolution (`.extractor.impl`)

### CODE REVIEW

File: engine/src/main/java/org/hibernate/search/v6poc/entity/model/spi/MappableTypeModel.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import org.hibernate.search.v6poc.entity.mapping.building.spi.IndexManagerBuildingState;
+import org.hibernate.search.v6poc.entity.mapping.building.spi.TypeMetadataContributorProvider;
+
+/**
+ * A representation of an entity type that can be mapped to an index.
+ *
+ * @see org.hibernate.search.v6poc.entity.mapping.building.spi.Mapper#addIndexed(MappableTypeModel, IndexManagerBuildingState, TypeMetadataContributorProvider)
+ */
+public interface MappableTypeModel {
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply engine/src/main/java/org/hibernate/search/v6poc/entity/model/spi/MappableTypeModel.java]: > Aren't you talking about IndexedTypeModel?

Not really, this also include types which are not annotated with `@Indexed`, but end up being embedded.

File: mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/extractor/builtin/CollectionElementExtractor.java [hunk 2]
```diff
@@ -4,19 +4,14 @@
  * License: GNU Lesser General Public License (LGPL), version 2.1 or later
  * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
  */
-package org.hibernate.search.v6poc.entity.pojo.extractor.impl;
+package org.hibernate.search.v6poc.entity.pojo.extractor.builtin;
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/extractor/builtin/CollectionElementExtractor.java]: Nope, users could need to reference those types when building a complex extractor chain, such as in the following case:

```
public class MyBean {
    @IndexedEmbedded(extractors = {
        @ContainerValueExtractorBeanReference(type = MyCustomContainerExtractor.class), 
        @ContainerValueExtractorBeanReference(type = CollectionElementExtractor.class)
    })
   private MyCustomContainer<Collection<String>> property;
} 
```

File: engine/src/main/java/org/hibernate/search/v6poc/entity/model/spi/MappableTypeModel.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (yrodiere) [reply engine/src/main/java/org/hibernate/search/v6poc/entity/model/spi/MappableTypeModel.java]: Also, some of the instances of `MappableTypeModel` could end up not having any associated mapping, for instance the type model for `Object.class`. So `MappedTypeModel` would be wrong too.

[the remark this answers is not in the document]
Comment (gsmet) [reply engine/src/main/java/org/hibernate/search/v6poc/entity/model/spi/MappableTypeModel.java]: Well, I was not talking about objects marked with `@Indexed` but generally of types that ends up in the index.

But OK.

**Comment by gsmet** [comment]
> About annotating type paramters, I don't like the idea for Hibernate Search for two reasons

I can understand your point of view. I don't particularly like your syntax but it has the advantage of having all the fields at the same level and it's probably easier to understand the structure of the index.

About your type parameter resolution algorithm, I will probably have to study it more closely at some point. We have a suboptimal piece of code for that in HV and I was surprised you could manage all these cases with your code, thus my question.

I think the supported cases are enough for now and it looks good so feel free to merge it.

**Comment by yrodiere** [comment]
Merged, thanks!

> About your type parameter resolution algorithm, I will probably have to study it more closely at some point. We have a suboptimal piece of code for that in HV and I was surprised you could manage all these cases with your code, thus my question.

In case you want to investigate later, some info:

* I didn't invent anything, I simply had a look at Guava's `TypeResolver` (`com.google.common.reflect.TypeResolver`) and deduced what was needed in our (very specific) case.
* The algorithm is in `org.hibernate.search.v6poc.entity.pojo.util.impl.GenericTypeContext` and is tested in `org.hibernate.search.v6poc.entity.pojo.util.impl.GenericTypeContextTest`
* The algorithm involves walking the type's inheritance tree and populating a map, so it may not meet HV's efficiency standards. I think we're generally more lenient in Search, since bootstrapping involves I/O operations which are generally an order of magnitude slower than anything else.

**Comment by gsmet** [comment]
> The algorithm involves walking the type's inheritance tree and populating a map

Yeah, we do that too and this part is particularly slow. The fact is that unfortunately, for HV, we have to do some of this at runtime as cascading consider the runtime type and not the original declared type.

**Comment by yrodiere** [comment]
> cascading consider the runtime type and not the original declared type.

Great that we need a schema for the index then: it basically prevents this kind of feature :]

**Pull request merged**